src/internet_identity/src/storage.rs

```diff
@@ -511,6 +539,97 @@ impl<M: Memory> Storage<M> {
+            },
+            5 => MigrationState::Finished,
+            version => trap(&format!("unexpected header version: {}", version)),
+        }
+    }
+
+    pub fn configure_migration(&mut self, migration_batch_size: u32) {
+        if self.header.version == 5 {
+            // migration is already done, nothing to do
+            return;
+        }
+
+        if self.header.version == 3 {
+            // only initialize this on the first migration configuration
+            self.header.version = 4;
+            self.header.new_layout_start = self.header.num_users;
+        }
+
+        self.header.migration_batch_size = migration_batch_size;
+        self.flush();
+    }
+
+    fn migrate_record_batch(&mut self) -> Result<(), StorageError> {
+        if self.header.version == 3 || self.header.version == 5 {
+            // migration is not started or already done, nothing to do
+            return Ok(());
+        }
+
+        for _ in 0..min(
+            self.header.new_layout_start,
+            self.header.migration_batch_size,
+        ) {
+            self.migrate_next_record()?;
+        }
+
+        if self.header.new_layout_start == 0 {
+            self.finalize_migration();
+        }
+        self.flush();
+        Ok(())
+    }
+
+    fn migrate_next_record(&mut self) -> Result<(), StorageError> {
+        let record_number = self.header.new_layout_start - 1;
+        let old_schema_bytes = self.read_entry_bytes(record_number);
+
+        if old_schema_bytes.len() == 0 {
+            // This anchor was only allocated but never written
+            // --> nothing to migrate just update pointer.
+            self.header.new_layout_start -= 1;
+            return Ok(());
+        }
+
+        let devices: Vec<DeviceDataInternal> =
+            candid::decode_one(&old_schema_bytes).map_err(StorageError::DeserializationError)?;
+        let anchor = Anchor {
+            devices: devices.into_iter().map(|d| Device::from(d)).collect(),
+        };
+
+        let new_schema_bytes =
+            candid::encode_one(anchor).map_err(StorageError::SerializationError)?;
+        self.write_entry_bytes(record_number, new_schema_bytes)?;
+        self.header.new_layout_start -= 1;
+        Ok(())
+    }
+
+    fn finalize_migration(&mut self) {
+        if self.header.version == 5 {
+            // migration is already done, nothing to do
+            return;
+        }
+
+        assert_eq!(
+            self.header.version, 4,
+            "unexpected header version during migration finalization {}",
+            self.header.version
+        );
+        assert_eq!({ self.header.new_layout_start }, 0, "cannot finalize migration when not all anchors were migrated! Remaining anchors to migrate: {}", { self.header.new_layout_start });
+
+        self.header.version = 5;
+        // clear now unused header fields
+        self.header.migration_batch_size = 0;
+        self.header.migration_batch_size = 0;
```

Comment:
`self.header.migration_batch_size = 0` is set twice here, is the 2nd one meant to be `self.header.new_layout_start = 0`?

`layout_start` is already 0 at this point anyway